util/dvsim/Deploy.py

```diff
@@ -548,7 +579,7 @@ def __init__(self, build_mode, sim_cfg):
         self.pass_patterns = []
         self.fail_patterns = []
 
-        self.mandatory_cmd_attrs = {
+        self.mandatory_cmd_attrs = OrderedDict({
```

Comment:
Why all the ordered dictionaries? I assume you're trying to force an ordering on the keys so that a comparison will go through (although I can't see which one). Wouldn't it be neater to do that at comparison time? (checking `set(a.keys()) == set(b.keys())` is the usual trick)

Reply:
The keys and values are recursively substituted to eventually resolve into a build command, so comparing the keys being equal would be imprecise. The OrderedDict is to ensure that the command is constructed the same way in all individual `build_mode`s so that the `is_equal()` function above works robustly.

Reply:
Right, understood. If I understand right, the one line change in PR #3125 would have the same effect, and I think it's a bit simpler. Could we do that instead?

Reply:
Looks like I would have to use sorted keys within `_parse_dict` as well - no longer a one line change...

Reply:
I changed the implementation to initialize the two vars in the base class as OrderedDict so the extended classes only update the dicts. This avoids the overhead of sorting the keys every time. Hope this is better.

Reply:
Hmm, ok. Well, I'm still not keen on this approach where we change the data structure throughout the program just to make it easier to ensure a projection is canonical. But you clearly want to do it this way, so never mind.

Reply:
Yes, I am changing the data structure, but now that's only at one place - in the base class' __init__ method. The extended classes do not need to know about it anymore. I thought that solved your original comment on OrderedDict proliferation, while also serving my need AND avoiding any extra sorting work. I am really interested in understanding your thoughts on why is this not better, maybe I am missing some point.

Reply:
Well, I guess it's a matter of preference. My take is that we're passing around some dictionaries where we fundamentally don't care about ordering, it's just that we want to compare the maps they define.

One way to do that is to fix an ordering, using `OrderedDict`. This is slightly conservative because two `OrderedDict` objects might define the same map from key to value, but not compare equal (if they were constructed in a different order from one another). I guess that doesn't happen with the current dvsim codebase, so that's not a big deal at the moment.

We "project down" from the dict object to something ordered in `Deploy.construct_cmd`. (Your comment above mentions `_parse_dict` too, but I don't think that's right: I don't think the order we iterate through the dictionary matters there.) So another approach is to force an ordering at that point. My proposal ordered alphabetically by key, but it doesn't really matter what ordering you choose, so long as it's consistent. This "resolves the ordering ambiguity" at exactly the place it matters: the only place where we go from a dict to a (list/string/something ordered).

I prefer the second approach (obviously! Otherwise I wouldn't have written the PR) and think it's cleaner not to inject structure into a program where it doesn't matter. Here, that "structure injection" is recording an ordering as part of the construction of a dictionary, when the program doesn't care what the ordering is, just that it is *something*.

You also mention the overhead of sorting keys every time. I suspect that the runtime is completely irrelevant here, but I wouldn't be so sure that sorting explicitly in `construct_cmd` is slower than using an `OrderedDict`. If you do the latter, then every dictionary operation you do on the object will be slightly slower than it would be on a plain `Dict`. No idea which version wins overall (and this is Python, so it doesn't really matter as long as it's not a factor of 100!), but I don't think it's completely clear-cut which option is faster.

Reply:
Oh, I should say: I am a definite fan of the change to use `SomeDictClass.update` in the subclass constructors (and call the superclass constructor), rather than only setting the class member there. It feels a little less "duck typed" to have the superclass define the shared member variables.

Reply:
Thanks for making your point clear! I did miss the fact about OrderedDict being probably slower than plain dict, speed wise it likely does not matter. While we do need the determinism, we do not need it be strictly in input order. I have now changed those to `dict`.